synapse/storage/databases/main/deviceinbox.py

```diff
@@ -136,6 +137,79 @@ def process_replication_rows(self, stream_name, instance_name, token, rows):
     def get_to_device_stream_token(self):
         return self._device_inbox_id_gen.get_current_token()
 
+    async def get_new_messages(
+        self,
+        user_ids: Collection[str],
+        from_stream_id: int,
+        to_stream_id: int,
+    ) -> Dict[Tuple[str, str], List[JsonDict]]:
+        """
+        Retrieve to-device messages for a given set of user IDs.
+
+        Only to-device messages with stream ids between the given boundaries
+        (from < X <= to) are returned.
+
+        Note that a stream ID can be shared by multiple copies of the same message with
+        different recipient devices. Each (device, message_content) tuple has their own
+        row in the device_inbox table.
+
+        Args:
+            user_ids: The users to retrieve to-device messages for.
+            from_stream_id: The lower boundary of stream id to filter with (exclusive).
+            to_stream_id: The upper boundary of stream id to filter with (inclusive).
+
+        Returns:
+            A list of to-device messages.
+        """
+        # Bail out if none of these users have any messages
+        for user_id in user_ids:
+            if self._device_inbox_stream_cache.has_entity_changed(
+                user_id, from_stream_id
+            ):
+                break
+        else:
+            return {}
+
+        def get_new_messages_txn(txn: LoggingTransaction):
+            # Build a query to select messages from any of the given users that are between
+            # the given stream id bounds
+
+            # Scope to only the given users. We need to use this method as doing so is
+            # different across database engines.
+            many_clause_sql, many_clause_args = make_in_list_sql_clause(
+                self.database_engine, "user_id", user_ids
+            )
+
+            sql = f"""
+                SELECT user_id, device_id, message_json FROM device_inbox
+                WHERE {many_clause_sql}
+                AND ? < stream_id AND stream_id <= ?
+                ORDER BY stream_id ASC
```

Comment:
have you checked what the optimiser does with this? I can't remember what indexes we have, but i suspect if you have more than one or two users, it's going to start table-scanning.

Reply:
Standard indexes for this table are:

```
synapse=> \d device_inbox
               Table "public.device_inbox"
    Column     |  Type  | Collation | Nullable | Default 
---------------+--------+-----------+----------+---------
 user_id       | text   |           | not null | 
 device_id     | text   |           | not null | 
 stream_id     | bigint |           | not null | 
 message_json  | text   |           | not null | 
 instance_name | text   |           |          | 
Indexes:
    "device_inbox_stream_id_user_id" btree (stream_id, user_id)
    "device_inbox_user_stream_id" btree (user_id, device_id, stream_id)

synapse=> 
```

It seems that adding device IDs to this query speed it up significantly, especially when scanning a large range of stream IDs (e.g if the AS has been offline for a while):

<details>

<summary>query with no device IDs specified</summary>

```
synapse=# EXPLAIN ANALYZE -- no device IDs specified
SELECT user_id, device_id, message_json FROM device_inbox
WHERE user_id in ('@andrewm:amorgan.xyz', '@testguest:amorgan.xyz', '@telegram_68411305:amorgan.xyz', '@telegram_11305:amorgan.xyz', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20')                                           
AND 1000000 < stream_id AND stream_id <= 1518046
ORDER BY stream_id ASC;
                                                                                            QUERY PLAN                                                                                             
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using device_inbox_stream_id on device_inbox  (cost=0.43..598059.96 rows=3197064 width=453) (actual time=0.581..10466.706 rows=3257282 loops=1)
   Index Cond: ((1000000 < stream_id) AND (stream_id <= 1518046))
   Filter: (user_id = ANY ('{@andrewm:amorgan.xyz,@testguest:amorgan.xyz,@telegram_68411305:amorgan.xyz,@telegram_11305:amorgan.xyz,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20}'::text[]))
   Rows Removed by Filter: 46549
 Planning time: 0.630 ms
 Execution time: 10654.986 ms
(6 rows)
```

</details>

The largest slowdown (on amorgan.xyz) looks to be down to the size of the stream ID range - where a range size this large (518046) is indeed quite slow. Throwing in some device IDs though:

<details>

<summary>specifying some device IDs</summary>

```
synapse=# EXPLAIN ANALYZE                                                      
SELECT user_id, device_id, message_json FROM device_inbox
WHERE user_id in ('@andrewm:amorgan.xyz', '@testguest:amorgan.xyz', '@telegram_68411305:amorgan.xyz', '@telegram_11305:amorgan.xyz', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20')
AND device_id in ('XFHILLDHIT', 'CJYNODEIGN', 'TKYYIPMJAN', 'WQWGGLVYVX', 'ODIEEXSYLA')
AND 1000000 < stream_id AND stream_id <= 1518046
ORDER BY stream_id ASC;

                                                                                                                                                                                QUERY PLAN                                                                                                                                                                                
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=27251.02..27275.14 rows=9649 width=453) (actual time=1595.313..1602.499 rows=31182 loops=1)
   Sort Key: stream_id
   Sort Method: external merge  Disk: 17824kB
   ->  Index Scan using device_inbox_user_stream_id on device_inbox  (cost=0.56..24631.44 rows=9649 width=453) (actual time=1.719..1563.038 rows=31182 loops=1)
         Index Cond: ((user_id = ANY ('{@andrewm:amorgan.xyz,@testguest:amorgan.xyz,@telegram_68411305:amorgan.xyz,@telegram_11305:amorgan.xyz,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20}'::text[])) AND (device_id = ANY ('{XFHILLDHIT,CJYNODEIGN,TKYYIPMJAN,WQWGGLVYVX,ODIEEXSYLA,VVDOQSPBZK}'::text[])) AND (1000000 < stream_id) AND (stream_id <= 1518046))
 Planning time: 0.194 ms
 Execution time: 1607.424 ms
(7 rows)
```

</details>

and we take advantage of our `device_inbox_user_stream_id` index.

In practice, the range of stream IDs could grow quite large in practice if the AS has been offline for a while. I'm not sure if the set of users/devices here would realistically get that large (unless a user had a *ton* of devices).

Reply:
Blocked on https://github.com/matrix-org/synapse/pull/11215/files#r752249314.

Reply:
`_get_device_messages` (implemented in https://github.com/matrix-org/synapse/commit/822e92a62f35f52a1cefe42d5af6bed75da93c62) queries using the `device_id` of the `device_inbox_user_stream_id`, resulting in a much faster query.